### PR TITLE
Implement strict portability import

### PR DESCRIPTION
**Pull request by sfmskywalker**
## Summary
- add target-state and atomic apply hooks to the portability store contract
- implement strict import planning, no-op reuse, divergent collision diagnostics, and write import execution
- add in-memory and SQLite provider import apply paths
- add strict import tests for empty stores, identical reuse, strict collision failure, and SQLite persistence

## Validation
- dotnet test Aster.sln
- dotnet build Aster.sln /m:1
- git diff --check
- dotnet format Aster.sln --verify-no-changes --verbosity minimal --include <changed files>

## Notes
- RemapDivergent identity mapping remains intentionally deferred to the next import slice.

### CODE REVIEW

File: src/core/Aster.Core/InMemory/InMemoryPortabilityStore.cs [hunk 1]
```diff
@@ -47,6 +47,90 @@ public ValueTask<PortableStoreSnapshot> ReadSnapshotAsync(
...
+        {
+            await resourceStore.SaveVersionAsync(resource, CancellationToken.None);
+        }
+
+        foreach (var state in plannedSnapshot.ActivationStates
+            .OrderBy(static state => state.ResourceId, StringComparer.Ordinal)
+            .ThenBy(static state => state.Channel, StringComparer.Ordinal))
+        {
+            await resourceStore.UpdateActivationAsync(state.ResourceId, state.Channel, state, CancellationToken.None);
+        }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/InMemory/InMemoryPortabilityStore.cs]: Fixed in eaf5446: ApplyImportAsync now checks the provided cancellation token between import items and passes it through to activation updates. Resource imports now use the in-memory import path directly, so they also respect the explicit cancellation checks.

File: src/core/Aster.Core/InMemory/InMemoryResourceDefinitionStore.cs [hunk 2]
```diff
@@ -120,6 +120,18 @@ internal IReadOnlyList<ResourceDefinition> GetDefinitionVersions(string definiti
...
+    /// <summary>
+    /// Imports an exact definition version.
+    /// </summary>
+    internal void ImportDefinitionVersion(ResourceDefinition definition)
+    {
+        ArgumentNullException.ThrowIfNull(definition);
+
+        var versions = definitions.GetOrAdd(definition.DefinitionId, _ => []);
+        lock (versions)
+            versions.Add(definition);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/InMemory/InMemoryResourceDefinitionStore.cs]: Fixed in eaf5446: imported definition versions are now inserted in version order and duplicate imported versions are rejected. I also added a regression test proving latest-definition lookup remains on v2 after importing v2 before v1.

File: src/core/Aster.Core/Services/ResourcePortabilityService.cs [hunk 3]
```diff
@@ -246,4 +437,60 @@ private static PortableDiagnostic InvalidExportScope(string path, string message
             Path = path,
             Message = message,
         };
+
+    private static bool ContentEquals<T>(T left, T right) =>
+        string.Equals(
+            JsonSerializer.Serialize(left, JsonOptions),
+            JsonSerializer.Serialize(right, JsonOptions),
+            StringComparison.Ordinal);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/Services/ResourcePortabilityService.cs]: Fixed in eaf5446: import planning now uses structural equality for definitions, resources, and activation state. Dictionary payloads are compared by key/content, and ActiveVersions is canonicalized as an unordered version set. Added a no-op regression test for reordered active versions.

File: src/core/Aster.Core/InMemory/InMemoryResourceDefinitionStore.cs [hunk 4]
```diff
@@ -120,6 +120,18 @@ internal IReadOnlyList<ResourceDefinition> GetDefinitionVersions(string definiti
...
+    }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/InMemory/InMemoryResourceDefinitionStore.cs:142]: Fixed in eaf5446: in-memory definition and resource imports now insert versions in sorted order and reject duplicates, preserving the versions[^1] latest-version invariant. Added an out-of-order import regression test covering both definition and resource latest lookup.

File: src/core/Aster.Core/Services/ResourcePortabilityService.cs [hunk 5]
```diff
@@ -238,6 +370,65 @@ private static List<PortableDiagnostic> ValidateSnapshot(PortableSnapshot snapsh
...
+                Path = path,
+                Message = message,
+            }
+            : new PortableDiagnostic
+            {
+                Code = PortableDiagnosticCodes.ImportNotImplemented,
+                Severity = PortableDiagnosticSeverity.Error,
+                Path = path,
+                Message = $"{message} RemapDivergent import is planned for the next import slice.",
+            };
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/Services/ResourcePortabilityService.cs:464]: Fixed in eaf5446: RemapDivergent now emits PortableDiagnosticCodes.RemapDivergentNotImplemented instead of the generic ImportNotImplemented code, with a regression test for that diagnostic.

File: test/Aster.Tests/SqliteJson/SqliteJsonPortabilityStoreTests.cs [hunk 6]
```diff
@@ -90,6 +90,28 @@ await definitionStore.RegisterDefinitionAsync(new ResourceDefinitionBuilder()
...
+    [Fact]
+    public async Task ImportAsync_WithSqlite_PersistsSnapshotForLaterExport()
+    {
+        await using var provider = CreateServiceProvider();
+        var portability = provider.GetRequiredService<IResourcePortabilityService>();
+        var snapshot = CreateSnapshot();
+
+        var import = await portability.ImportAsync(snapshot);
+        var export = await portability.ExportAsync(new PortableSnapshotExportRequest
+        {
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply test/Aster.Tests/SqliteJson/SqliteJsonPortabilityStoreTests.cs]: Fixed in ae21c10: the SQLite import test now disposes the import provider and creates a fresh provider against the same database path before exporting, so it verifies persistence across provider lifetimes. Validation: dotnet test Aster.sln --no-restore, dotnet build Aster.sln /m:1, dotnet format --verify-no-changes, and git diff --check.

File: src/core/Aster.Core/Services/ResourcePortabilityService.cs [hunk 7]
```diff
@@ -238,6 +371,65 @@ private static List<PortableDiagnostic> ValidateSnapshot(PortableSnapshot snapsh
...
+            snapshot.Definitions.Select(static definition => DefinitionVersionId(definition)),
+            "definitions",
+            "Snapshot contains duplicate definition version identities."));
+        diagnostics.AddRange(FindDuplicateKeys(
+            snapshot.Resources.Select(static resource => ResourceVersionId(resource)),
+            "resources",
+            "Snapshot contains duplicate resource version identities."));
+        diagnostics.AddRange(FindDuplicateKeys(
+            snapshot.Resources.Select(static resource => resource.Id),
+            "resources",
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/Services/ResourcePortabilityService.cs]: Fixed in a91c234: duplicate version-specific resource IDs now report under resources/id/{id}, making the field identity explicit. Added a regression test for the diagnostic path. Validation: dotnet test Aster.sln --no-restore, dotnet build Aster.sln /m:1, dotnet format --verify-no-changes, and git diff --check.

**Comment by sfmskywalker** [comment]
Addressed the Greptile 4/5 finding in 73ac370: the in-memory ApplyImportAsync path now records applied definitions, resources, and activation states and rolls them back if any apply step fails or cancellation interrupts before completion. Added a direct rollback regression test. Validation: dotnet test Aster.sln --no-restore, dotnet build Aster.sln /m:1, dotnet format --verify-no-changes, and git diff --check.

File: src/core/Aster.Core/InMemory/InMemoryResourceDefinitionStore.cs [hunk 8]
```diff
@@ -120,6 +120,27 @@ internal IReadOnlyList<ResourceDefinition> GetDefinitionVersions(string definiti
...
     }
 
+    /// <summary>
+    /// Imports an exact definition version.
+    /// </summary>
+    internal void ImportDefinitionVersion(ResourceDefinition definition)
+    {
+        ArgumentNullException.ThrowIfNull(definition);
+
+        var versions = definitions.GetOrAdd(definition.DefinitionId, _ => []);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/InMemory/InMemoryResourceDefinitionStore.cs:130]: Fixed in 8466b5b: RegisterDefinitionAsync now derives the next version from the current highest ordered version instead of versions.Count + 1. Added a regression test covering registration after importing v10.

File: src/core/Aster.Core/InMemory/InMemoryResourceStore.cs [hunk 9]
```diff
@@ -45,6 +45,79 @@ public sealed class InMemoryResourceStore : IResourceVersionReader, IResourceVer
...
+            else
+                channelActivations[channel] = state.ActiveVersions.ToHashSet();
+        }
+
+        var states = ActivationStates.GetOrAdd(
+            resourceId,
+            _ => new ConcurrentDictionary<string, ActivationState>(StringComparer.Ordinal));
+
+        if (state is null)
+            states.Remove(channel, out _);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/InMemory/InMemoryResourceStore.cs]: Fixed in 8466b5b: RestoreActivationState now uses ConcurrentDictionary.TryRemove for activation-state rollback removal. Validation passes across test/build/format.

File: src/core/Aster.Core/InMemory/InMemoryPortabilityStore.cs [hunk 10]
```diff
@@ -47,6 +47,124 @@ public ValueTask<PortableStoreSnapshot> ReadSnapshotAsync(
...
+                appliedResources.Add(resource);
+            }
+
+            foreach (var state in plannedSnapshot.ActivationStates
+                .OrderBy(static state => state.ResourceId, StringComparer.Ordinal)
+                .ThenBy(static state => state.Channel, StringComparer.Ordinal))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                var previous = resourceStore.GetActivationState(state.ResourceId, state.Channel);
+                await resourceStore.UpdateActivationAsync(state.ResourceId, state.Channel, state, CancellationToken.None);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/InMemory/InMemoryPortabilityStore.cs]: Fixed in 8466b5b: ApplyImportAsync now passes the caller cancellation token through to UpdateActivationAsync. The rollback path added in 73ac370 handles cancellation observed after earlier writes.

File: src/core/Aster.Core/Services/ResourcePortabilityService.cs [hunk 11]
```diff
@@ -246,4 +438,114 @@ private static PortableDiagnostic InvalidExportScope(string path, string message
...
+
+        foreach (var (key, leftValue) in left.OrderBy(static pair => pair.Key, StringComparer.Ordinal))
+        {
+            if (!right.TryGetValue(key, out var rightValue))
+                return false;
+
+            if (!JsonContentEquals(leftValue, rightValue))
+                return false;
+        }
+
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/Services/ResourcePortabilityService.cs:523]: Fixed in 4e5c6eb: divergent collisions now emit identity-map entries with PortableIdentityMappingReason.CollidedDivergent for definitions, resources, and activation entries. Added regression coverage for strict divergent definition collisions and updated the contract enum. Validation: dotnet test Aster.sln --no-restore, dotnet build Aster.sln /m:1, dotnet format --verify-no-changes, and git diff --check.

File: src/core/Aster.Core/Services/ResourcePortabilityService.cs [hunk 12]
```diff
@@ -246,4 +441,123 @@ private static PortableDiagnostic InvalidExportScope(string path, string message
...
+        && left.Created == right.Created
+        && string.Equals(left.Owner, right.Owner, StringComparison.Ordinal)
+        && string.Equals(left.Hash, right.Hash, StringComparison.Ordinal)
+        && DictionaryContentEquals(left.Aspects, right.Aspects);
+
+    private static bool ContentEquals(ActivationState left, ActivationState right) =>
+        string.Equals(left.ResourceId, right.ResourceId, StringComparison.Ordinal)
+        && string.Equals(left.Channel, right.Channel, StringComparison.Ordinal)
+        && left.LastUpdated == right.LastUpdated
+        && left.ActiveVersions.Order().SequenceEqual(right.ActiveVersions.Order());
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/Services/ResourcePortabilityService.cs]: Fixed in 4fdac5a: activation equality now treats ActiveVersions as a logical set by applying Distinct() before ordering/comparison. Extended the no-op regression test to cover duplicate active-version values. Validation: dotnet test Aster.sln --no-restore, dotnet build Aster.sln /m:1, dotnet format --verify-no-changes, and git diff --check.

File: src/core/Aster.Core/Services/ResourcePortabilityService.cs [hunk 13]
```diff
@@ -246,4 +441,123 @@ private static PortableDiagnostic InvalidExportScope(string path, string message
...
+
+    private static bool ContentEquals(ResourceDefinition left, ResourceDefinition right) =>
+        string.Equals(left.DefinitionId, right.DefinitionId, StringComparison.Ordinal)
+        && string.Equals(left.Id, right.Id, StringComparison.Ordinal)
+        && left.Version == right.Version
+        && left.IsSingleton == right.IsSingleton
+        && DictionaryContentEquals(left.AspectDefinitions, right.AspectDefinitions);
+
+    private static bool ContentEquals(Resource left, Resource right) =>
+        string.Equals(left.ResourceId, right.ResourceId, StringComparison.Ordinal)
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/Services/ResourcePortabilityService.cs:492]: Fixed in 0fdb228: ImportAsync now catches non-cancellation provider apply failures and returns PortableImportStatus.Failed with an import-apply-failed diagnostic while preserving the planned identity map. Added a service-level regression test with a throwing portability store. Validation: dotnet test Aster.sln --no-restore, dotnet build Aster.sln /m:1, dotnet format --verify-no-changes, and git diff --check.

File: src/core/Aster.Core/Services/ResourcePortabilityService.cs [hunk 14]
```diff
@@ -78,53 +83,184 @@ public ValueTask<PortableSnapshotValidationResult> ValidateAsync(
...
-                    Severity = PortableDiagnosticSeverity.Error,
-                    Message = "Snapshot import is not implemented in this export-focused slice.",
-                },
-            ],
-        });
+            var key = (definition.DefinitionId, definition.Version);
+            if (!existingDefinitions.TryGetValue(key, out var existing))
+            {
+                plannedDefinitions.Add(definition);
+                identityMap.Add(Preserved(PortableEntityKind.DefinitionVersion, DefinitionVersionId(definition)));
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/Services/ResourcePortabilityService.cs:189]: Fixed in fabe1b6: BuildImportPlanAsync now checks cancellation inside the definitions, resources, and activation-state planning loops.

File: src/persistence/Aster.Persistence.SqliteJson/SqliteJsonResourceStore.cs [hunk 15]
```diff
@@ -352,6 +418,33 @@ private async Task<List<ResourceDefinition>> ReadScopedDefinitionVersionsAsync(
...
+        var definitions = await ReadPayloadsByTextIdsAsync<ResourceDefinition>(
+            tableName: "resource_definitions",
+            columnName: "definition_id",
+            ids: definitionIds,
+            orderBy: "definition_id, version",
+            cancellationToken);
+
+        return definitions
+            .Where(definition => definitionVersions.Contains((definition.DefinitionId, definition.Version)))
+            .ToList();
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/persistence/Aster.Persistence.SqliteJson/SqliteJsonResourceStore.cs]: Fixed in fabe1b6: SQLite target-state definition reads now query exact (definition_id, version) pairs in bounded batches instead of loading every version for each definition id.

File: src/persistence/Aster.Persistence.SqliteJson/SqliteJsonResourceStore.cs [hunk 16]
```diff
@@ -256,6 +256,72 @@ public async ValueTask<PortableStoreSnapshot> ReadSnapshotAsync(
...
+            .ToHashSet();
+
+        return new PortableTargetState
+        {
+            Definitions = definitions,
+            Resources = resources,
+            ActivationStates = activationStates
+                .Where(state => activationKeys.Contains((state.ResourceId, state.Channel)))
+                .ToList(),
+        };
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/persistence/Aster.Persistence.SqliteJson/SqliteJsonResourceStore.cs:283]: Fixed in fabe1b6: SQLite target-state activation reads now query exact (resource_id, channel) pairs in bounded batches instead of loading all channels for each resource id.

File: src/core/Aster.Core/Services/ResourcePortabilityService.cs [hunk 17]
```diff
@@ -246,4 +465,123 @@ private static PortableDiagnostic InvalidExportScope(string path, string message
...
+        };
+
+    private static string DefinitionVersionId(ResourceDefinition definition) =>
+        $"{definition.DefinitionId}:{definition.Version}";
+
+    private static string ResourceVersionId(Resource resource) =>
+        $"{resource.ResourceId}:{resource.Version}";
+
+    private static string ActivationEntryId(ActivationState state) =>
+        $"{state.ResourceId}:{state.Channel}";
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/Services/ResourcePortabilityService.cs]: Fixed in fabe1b6: portable identity-map keys now use JSON-array encoding for composite identities instead of colon-delimited strings, avoiding ambiguity when caller-supplied ids contain colons. Updated the strict collision regression expectation. Validation: dotnet test Aster.sln --no-restore, dotnet build Aster.sln /m:1, dotnet format --verify-no-changes, and git diff --check.

File: test/Aster.Tests/Portability/PortabilityImportTests.cs [hunk 18]
```diff
@@ -0,0 +1,386 @@
...
+        {
+            ScopeMode = PortableExportScopeMode.SelectedResources,
+            ResourceIds = ["product-1"],
+            ResourceVersionScope = PortableResourceVersionScope.AllVersions,
+        });
+
+        Assert.NotNull(exported.Snapshot);
+        Assert.Equal(snapshot.Definitions.Select(static definition => (definition.DefinitionId, definition.Id, definition.Version)), exported.Snapshot.Definitions.Select(static definition => (definition.DefinitionId, definition.Id, definition.Version)));
+        Assert.Equal(snapshot.Resources.Select(static resource => (resource.ResourceId, resource.Id, resource.DefinitionId, resource.DefinitionVersion, resource.Version)), exported.Snapshot.Resources.Select(static resource => (resource.ResourceId, resource.Id, resource.DefinitionId, resource.DefinitionVersion, resource.Version)));
+        Assert.Equal(snapshot.ActivationStates.Select(static state => (state.ResourceId, state.Channel, state.LastUpdated, Versions: state.ActiveVersions.ToArray())), exported.Snapshot.ActivationStates.Select(static state => (state.ResourceId, state.Channel, state.LastUpdated, Versions: state.ActiveVersions.ToArray())));
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply test/Aster.Tests/Portability/PortabilityImportTests.cs]: Fixed in aa75a2c: the in-memory portability test now compares active-version values using an order-stable string representation instead of array reference equality.

File: test/Aster.Tests/SqliteJson/SqliteJsonPortabilityStoreTests.cs [hunk 19]
```diff
@@ -90,6 +90,35 @@ await definitionStore.RegisterDefinitionAsync(new ResourceDefinitionBuilder()
...
+            ResourceIds = ["sqlite-product-1"],
+            ResourceVersionScope = PortableResourceVersionScope.AllVersions,
+        });
+
+        Assert.Equal(PortableImportStatus.Imported, import.Status);
+        Assert.NotNull(export.Snapshot);
+        Assert.Equal(snapshot.Definitions.Select(static definition => (definition.DefinitionId, definition.Id, definition.Version)), export.Snapshot.Definitions.Select(static definition => (definition.DefinitionId, definition.Id, definition.Version)));
+        Assert.Equal(snapshot.Resources.Select(static resource => (resource.ResourceId, resource.Id, resource.DefinitionId, resource.DefinitionVersion, resource.Version)), export.Snapshot.Resources.Select(static resource => (resource.ResourceId, resource.Id, resource.DefinitionId, resource.DefinitionVersion, resource.Version)));
+        Assert.Equal(snapshot.ActivationStates.Select(static state => (state.ResourceId, state.Channel, state.LastUpdated, Versions: state.ActiveVersions.ToArray())), export.Snapshot.ActivationStates.Select(static state => (state.ResourceId, state.Channel, state.LastUpdated, Versions: state.ActiveVersions.ToArray())));
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply test/Aster.Tests/SqliteJson/SqliteJsonPortabilityStoreTests.cs]: Fixed in aa75a2c: the SQLite portability test now compares active-version values using an order-stable string representation instead of array reference equality. Validation: dotnet test Aster.sln --no-restore, dotnet build Aster.sln /m:1, dotnet format --verify-no-changes, and git diff --check.

File: src/core/Aster.Core/Services/ResourcePortabilityService.cs [hunk 20]
```diff
@@ -246,4 +471,123 @@ private static PortableDiagnostic InvalidExportScope(string path, string message
...
+
+        return true;
+    }
+
+    private static bool JsonContentEquals<T>(T left, T right)
+    {
+        var leftNode = JsonSerializer.SerializeToNode(left, JsonOptions);
+        var rightNode = JsonSerializer.SerializeToNode(right, JsonOptions);
+
+        return JsonNode.DeepEquals(leftNode, rightNode);
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply src/core/Aster.Core/Services/ResourcePortabilityService.cs]: Fixed in 41a3cae: JsonContentEquals now catches JsonException/NotSupportedException from JSON serialization and treats the values as non-equal, so import planning returns a divergent-collision diagnostic instead of throwing. Added a regression test covering an unsupported CLR aspect value. Validation: dotnet test Aster.sln --no-restore, dotnet build Aster.sln /m:1, dotnet format --verify-no-changes, and git diff --check.

File: test/Aster.Tests/Portability/PortabilityImportTests.cs [hunk 21]
```diff
@@ -0,0 +1,417 @@
...
+                },
+            ],
+        };
+
+        var result = await portability.PreviewImportAsync(unsupportedSnapshot);
+
+        Assert.False(result.CanImport);
+        var diagnostic = Assert.Single(result.Diagnostics);
+        Assert.Equal(PortableDiagnosticCodes.DivergentIdentityCollision, diagnostic.Code);
+    }
```
[the remark this answers is not in the document]
Comment (sfmskywalker) [reply test/Aster.Tests/Portability/PortabilityImportTests.cs:181]: Fixed in 360da5c: the regression setup now imports a resource with the same aspect key using a serializable value, then previews the same key with an unsupported CLR value. This forces JsonContentEquals through the serialization path and would fail without the exception handling added in 41a3cae. Validation: dotnet test Aster.sln --no-restore, dotnet build Aster.sln /m:1, dotnet format --verify-no-changes, and git diff --check.
